library/ssl_tls12_server.c

```diff
@@ -2531,10 +2531,14 @@ static int ssl_write_certificate_request( mbedtls_ssl_context *ssl )
         if( ! mbedtls_ssl_sig_alg_is_supported( ssl, *sig_alg ) )
             continue;
 
-        MBEDTLS_PUT_UINT16_BE( *sig_alg, p, sa_len );
+        /* Pre-increment so as to leave space for
+         * supported_signature_algorithms to be written. */
```

Comment:
Actually this comment is wrong: the increment is to leave space for the _current entry_, not for the whole `supported_signature_algorithms`. Incrementing `sa_len` creates enough space for one entry, then `PUT(…, sa_len)` writes one entry in the newly created space.

Reply:
Ok, misread the diagram. Can we agree on calling this the 'list length' then?

Reply:
Yes, but the connection between list length and preincrementation is not obvious, and “pre-increment so as to leave space for the list length” is not something I'd understand without working it backwards from the code. How about this:

> Write elements at offsets starting from 1 (offset 0 is for the length). Thus the offset of each element is the length of the partial list including the element.